datafusion/functions/src/core/named_struct.rs

```diff
@@ -203,12 +137,19 @@ impl ScalarUDFImpl for NamedStructFunc {
         ))))
     }
 
-    fn invoke_batch(
```

Comment:
If delete `invoke_batch` function, the default `invoke_batch` function will call `invoke` when args is not empty, but `invoke` function is left impl be Specific type, is it ok?

Reply:
I wasn't sure what the intention is for the deprecated functions from an API user point of view. I had interpreted ScalarUDF(Impl) as an extension point rather than as an API library users would call directly. If that assumption is correct, then it doesn't matter if we remove `invoke_batch` since `ScalarFunctionExpr` never calls this method. The only usages of `invoke_batch` are in benchmark and test code.
Perhaps @alamb, who logged #13515, can provide some insight here.

Reply:
THe reason we kept around deprecated functions is to give downstream users of DataFusion time to adjust their code -- especially on upgrade having a deprecated function with guidance of what to change has been helpful

You can read more about this strategy here:
- https://datafusion.apache.org/library-user-guide/api-health.html

Reply:
I'm coming from the Java world so I'll use the terminology from there; not sure what the equivalent is in Rust lingo. It's not explicitly stated which parts of the library are [Service Provider Interface](https://en.wikipedia.org/wiki/Service_provider_interface) vs [Application Programming Interface](https://en.wikipedia.org/wiki/API). My assumption was that `ScalarUDFImpl` is SPI and `invoke_batch` was kept around to not break all existing implementations and `ScalarFunctionExpr` is the API side of things which doesn't expose `invoke_batch`. Is that a correct interpretation?

Reply:
> My assumption was that ScalarUDFImpl is SPI 

Yes, that is accurate in my understanding

> and invoke_batch was kept around to not break all existing implementations 

Yes that is also my understanding


> and ScalarFunctionExpr is the API side of things which doesn't expose invoke_batch. Is that a correct interpretation?

I would say `ScalarFunctionExpr` is an implementation detail of how functions are invoked in the `ExecutionPlan` (aka the physical execution)

The split between logical/physical plans is explained a bit in the API docs / intro videos in case you are interested:
https://docs.rs/datafusion/latest/datafusion/index.html#query-planning-and-execution-overview